harvester/metadata.py

```diff
@@ -0,0 +1,278 @@
+                continue
+            if content := og_tag.get("content"):  # type: ignore[union-attr]
+                content_stripped = content.strip()  # type: ignore[union-attr]
+                if content_stripped != "":
+                    og_tag_name_friendly = og_tag_name.replace(":", "_")
+                    tags[og_tag_name_friendly] = content_stripped
+        return tags
+
+    def _remove_fulltext_whitespace(self, fulltext: str) -> str:
+        """Cleanup fulltext as provided by Browsertrix crawl."""
+        fulltext = fulltext.replace("\n", " ")
+        fulltext = fulltext.replace("\r", " ")
+        fulltext = fulltext.replace("\t", " ")
+        # ruff: noqa: RET504
+        return fulltext
+
+    def _generate_fulltext_keywords(self, fulltext: str) -> str:
+        """Parse keywords from fulltext, using YAKE keyword extractor.
+
+        The yake method extract_keywords() returns a keyword and score of that keyword
+        from the fulltext as a tuple.  The score is not currently used.
+        """
+        keywords = self.keyword_extractor.extract_keywords(fulltext)
+        return ",".join([keyword for keyword, _score in keywords])
+
+    def parse_fulltext_fields(
+        self,
+        raw_fulltext: str | None,
+        # ruff: noqa: FBT001, FBT002
+        include_fulltext: bool = False,
+        include_fulltext_keywords: bool = True,
+    ) -> dict:
+        """Parse fulltext fields for output metadata."""
+        if not raw_fulltext:
+            return {
+                "fulltext": None,
+                "fulltext_50_words": None,
+                "fulltext_keywords": None,
+            }
+
+        fulltext = self._remove_fulltext_whitespace(raw_fulltext)
+        return {
+            "fulltext": fulltext if include_fulltext else None,
+            "fulltext_keywords": self._generate_fulltext_keywords(fulltext)
+            if include_fulltext_keywords
+            else None,
+        }
+
+    @property
+    def keyword_extractor(self) -> KeywordExtractor:
+        """Property to lazy load an instance of the yake KeywordExtract class.
+
+        This is a somewhat experimental feature, to extract keywords from a website's
+        full-text.  It may be determined that a more refined approach is needed, but for
+        now this provides a serviceable (if somewhat naive) list of keywords for websites.
+        """
+        if self._keyword_extractor is None:
+            # init keyword extractor
+            self._keyword_extractor = KeywordExtractor()
+
+            # update stopwords
+            self._keyword_extractor.stopword_set.update(
+                [word.lower().strip() for word in self.FULLTEXT_KEYWORD_STOPWORDS]
+            )
+        return self._keyword_extractor
+
+
+class CrawlMetadataRecords:
+    """Class to represent and serialize structured metadata extracted from crawl."""
+
+    def __init__(self, metadata_df: pd.DataFrame) -> None:
+        self.df = metadata_df
```

Comment:
Still would prefer `dataframe` over `df` but it's not the end of the world

Reply:
I think `_df` is a common pandas convention, but will update to `metadata_df`.